File: src/Urql.Core/Syntax/Templates/InterpolationTemplateParser.cs
using System.Text;
using Urql.Core.Diagnostics;

namespace Urql.Core.Syntax.Templates;

public static class InterpolationTemplateParser
{
    private sealed class Frame
    {
        public required int OpenIndex { get; init; }
        public List<TemplatePart> Parts { get; } = [];
        public StringBuilder Literal { get; } = new();
    }

    public static InterpolationTemplate Parse(
        string input,
        DiagnosticBag diagnostics,
        out bool hadInterpolation)
    {
        hadInterpolation = false;
        var positionMap = BuildPositionMap(input);

        var stack = new Stack<Frame>();
        stack.Push(new Frame { OpenIndex = -1 });

        for (var i = 0; i < input.Length; i++)
        {
            var c = input[i];
            var current = stack.Peek();

            if (c == '#')
            {
                hadInterpolation = true;
                FlushLiteral(current);
                stack.Push(new Frame { OpenIndex = i });
                continue;
            }

            if (c == '$')
            {
                if (stack.Count == 1)
                {
                    diagnostics.ReportError(
                        DiagnosticCode.UnexpectedCharacter,
                        "Unmatched interpolation close delimiter '$'.",
                        SpanAt(positionMap, i, i));
                    current.Literal.Append(c);
                    continue;
                }

                hadInterpolation = true;
                var finished = stack.Pop();
                FlushLiteral(finished);
                var part = new EmbeddedExpressionPart(
                    new InterpolationTemplate(finished.Parts),
                    SpanAt(positionMap, finished.OpenIndex, i));
                stack.Peek().Parts.Add(part);
                continue;
            }

            current.Literal.Append(c);
        }

        while (stack.Count > 1)
        {
            var unclosed = stack.Pop();
            diagnostics.ReportError(
                DiagnosticCode.UnexpectedCharacter,
                "Unmatched interpolation open delimiter '#'.",
                SpanAt(positionMap, unclosed.OpenIndex, unclosed.OpenIndex));
        }

        var root = stack.Pop();
        FlushLiteral(root);
        return new InterpolationTemplate(root.Parts);
    }

    private static void FlushLiteral(Frame frame)
    {
        if (frame.Literal.Length == 0)
        {
            return;
        }

        frame.Parts.Add(new LiteralTextPart(frame.Literal.ToString()));
        frame.Literal.Clear();
    }

    private static SourceSpan SpanAt(IReadOnlyList<SourcePosition> positions, int startIndex, int endIndex)
    {
        var start = startIndex >= 0 && startIndex < positions.Count ? positions[startIndex] : new SourcePosition(1, 1);
        var end = endIndex >= 0 && endIndex < positions.Count ? positions[endIndex] : start;
        return new SourceSpan(start, end);
    }

    private static List<SourcePosition> BuildPositionMap(string input)
    {
        var map = new List<SourcePosition>(input.Length);
        var line = 1;
        var column = 1;
        foreach (var c in input)
        {
            map.Add(new SourcePosition(line, column));
            if (c == '\n')
            {
                line++;
                column = 1;
            }
            else
            {
                column++;
            }
        }

        return map;
    }
}

File: web/packages/core/src/interpolation-template-parser.ts
import { DiagnosticBag, diagnosticCode } from "./diagnostics.js";
import { sourcePosition, sourceSpan, type SourcePosition, type SourceSpan } from "./source.js";
import type {
  EmbeddedExpressionPart,
  InterpolationTemplate,
  LiteralTextPart,
  TemplatePart
} from "./interpolation-template.js";

interface Frame {
  openIndex: number;
  parts: TemplatePart[];
  literal: string[];
}

export class InterpolationTemplateParser {
  static parse(
    input: string,
    diagnostics: DiagnosticBag,
    out: { hadInterpolation: boolean }
  ): InterpolationTemplate {
    out.hadInterpolation = false;
    const positionMap = buildPositionMap(input);
    const stack: Frame[] = [{ openIndex: -1, parts: [], literal: [] }];

    for (let index = 0; index < input.length; index += 1) {
      const character = input[index]!;
      const current = stack[stack.length - 1]!;

      if (character === "#") {
        out.hadInterpolation = true;
        flushLiteral(current);
        stack.push({ openIndex: index, parts: [], literal: [] });
        continue;
      }

      if (character === "$") {
        if (stack.length === 1) {
          diagnostics.reportError(
            diagnosticCode.unexpectedCharacter,
            "Unmatched interpolation close delimiter '$'.",
            spanAt(positionMap, index, index)
          );
          current.literal.push(character);
          continue;
        }

        out.hadInterpolation = true;
        const finished = stack.pop()!;
        flushLiteral(finished);
        const part: EmbeddedExpressionPart = {
          kind: "EmbeddedExpressionPart",
          content: { parts: finished.parts },
          span: spanAt(positionMap, finished.openIndex, index)
        };
        stack[stack.length - 1]!.parts.push(part);
        continue;
      }

      current.literal.push(character);
    }

    while (stack.length > 1) {
      const unclosed = stack.pop()!;
      diagnostics.reportError(
        diagnosticCode.unexpectedCharacter,
        "Unmatched interpolation open delimiter '#'.",
        spanAt(positionMap, unclosed.openIndex, unclosed.openIndex)
      );
    }

    const root = stack.pop()!;
    flushLiteral(root);
    return { parts: root.parts };
  }
}

function flushLiteral(frame: Frame): void {
  if (frame.literal.length === 0) {
    return;
  }

  const part: LiteralTextPart = {
    kind: "LiteralTextPart",
    text: frame.literal.join("")
  };
  frame.parts.push(part);
  frame.literal.length = 0;
}

function spanAt(positions: SourcePosition[], startIndex: number, endIndex: number): SourceSpan {
  const start =
    startIndex >= 0 && startIndex < positions.length ? positions[startIndex]! : sourcePosition(1, 1);
  const end = endIndex >= 0 && endIndex < positions.length ? positions[endIndex]! : start;
  return sourceSpan(start, end);
}

function buildPositionMap(input: string): SourcePosition[] {
  const map: SourcePosition[] = [];
  let line = 1;
  let column = 1;

  for (const character of input) {
    map.push(sourcePosition(line, column));
    if (character === "\n") {
      line += 1;
      column = 1;
    } else {
      column += 1;
    }
  }

  return map;
}
